core/src/test/java/org/apache/calcite/test/HepPlannerTest.java

```diff
@@ -251,14 +251,15 @@ private void assertIncludesExactlyOnce(String message, String digest, String sub
     subprogramBuilder.addMatchOrder(HepMatchOrder.TOP_DOWN);
     subprogramBuilder.addMatchLimit(1);
     subprogramBuilder.addRuleInstance(ProjectToCalcRule.INSTANCE);
+    subprogramBuilder.addRuleInstance(FilterToCalcRule.INSTANCE);
     subprogramBuilder.addRuleInstance(CalcMergeRule.INSTANCE);
 
     HepProgramBuilder programBuilder = HepProgram.builder();
     programBuilder.addSubprogram(subprogramBuilder.build());
 
     checkPlanning(
         programBuilder.build(),
-        "select upper(ename) from (select lower(ename) as ename from emp)");
+        "select upper(ename) from (select lower(ename) as ename from emp where empno = 100)");
```

Comment:
I mostly +1 on this, but we better have a way to detect that the subProgram really fired 2 times, or we still cannot find the bug if someone breaks the rule and this test fire 1 times but still got passed.

Reply:
You're right. So I changed this test and it will break if the subProgram does not really fire 2 times. The original plan of the test shows the difference.